unit_tests/log_transfer/test_main_chkpt_info.cpp

```diff
@@ -0,0 +1,800 @@
+mvcc_active_tran::~mvcc_active_tran () = default;
+mvcc_trans_status::mvcc_trans_status () = default;
+mvcc_trans_status::~mvcc_trans_status () = default;
+mvcctable::mvcctable () = default;
+mvcctable::~mvcctable () = default;
+
+log_append_info::log_append_info () = default;
+log_prior_lsa_info::log_prior_lsa_info () = default;
+
+int
+csect_enter (THREAD_ENTRY *thread_p, int cs_index, int wait_secs)
+{
+  assert (true);
+}
+
+void
+log_system_tdes::map_all_tdes (const map_func &func)
+{
+  std::lock_guard<std::mutex> lg (systb_Mutex);
+  for (auto &el : systb_System_tdes)
+    {
+      log_tdes *tdes = el.second;
+      assert (tdes != NULL);
+      func (*tdes);
+    }
+}
+#define NUM_TOTAL_TRAN_INDICES log_Gl.trantable.num_total_indices
+
+void
+logtb_clear_tdes (THREAD_ENTRY *thread_p, LOG_TDES *tdes)
+{
+  int i, j;
+  DB_VALUE *dbval;
+  HL_HEAPID save_heap_id;
+
+  tdes->isloose_end = false;
+  tdes->state = TRAN_ACTIVE;
+  LSA_SET_NULL (&tdes->head_lsa);
+  LSA_SET_NULL (&tdes->tail_lsa);
+  LSA_SET_NULL (&tdes->undo_nxlsa);
+  LSA_SET_NULL (&tdes->posp_nxlsa);
+  LSA_SET_NULL (&tdes->savept_lsa);
+  LSA_SET_NULL (&tdes->topop_lsa);
+  LSA_SET_NULL (&tdes->tail_topresult_lsa);
+  tdes->topops.last = -1;
+  tdes->gtrid = LOG_2PC_NULL_GTRID;
+  tdes->gtrinfo.info_length = 0;
+  tdes->cur_repl_record = 0;
+  tdes->append_repl_recidx = -1;
+  tdes->fl_mark_repl_recidx = -1;
+  LSA_SET_NULL (&tdes->repl_insert_lsa);
+  LSA_SET_NULL (&tdes->repl_update_lsa);
+  tdes->first_save_entry = NULL;
+  tdes->query_timeout = 0;
+  tdes->query_start_time = 0;
+  tdes->tran_start_time = 0;
+  tdes->waiting_for_res = NULL;
+  tdes->tran_abort_reason = TRAN_NORMAL;
+  tdes->num_exec_queries = 0;
+  tdes->suppress_replication = 0;
+  tdes->has_deadlock_priority = false;
+
+  tdes->num_log_records_written = 0;
+
+  LSA_SET_NULL (&tdes->rcv.tran_start_postpone_lsa);
+  LSA_SET_NULL (&tdes->rcv.sysop_start_postpone_lsa);
+  LSA_SET_NULL (&tdes->rcv.atomic_sysop_start_lsa);
+  LSA_SET_NULL (&tdes->rcv.analysis_last_aborted_sysop_lsa);
+  LSA_SET_NULL (&tdes->rcv.analysis_last_aborted_sysop_start_lsa);
+}
+
+log_tdes *
+systdes_create_tdes ()
+{
+  log_tdes *tdes = new log_tdes ();
+  return tdes;
+}
+
+log_tdes *
+log_system_tdes::rv_get_tdes (TRANID trid)
+{
+  auto it = systb_System_tdes.find (trid);
+  if (it != systb_System_tdes.end ())
+    {
+      return it->second;
+    }
+  else
+    {
+      return NULL;
+    }
+}
+
+LOG_TDES *
+logtb_rv_find_allocate_tran_index (THREAD_ENTRY *thread_p, TRANID trid, const LOG_LSA *log_lsa)
+{
+  LOG_TDES *tdes = NULL;		/* Transaction descriptor */
+
+  assert (trid != NULL_TRANID);
+
+  for (int i = 1; i < NUM_TOTAL_TRAN_INDICES; i++)
+    {
+      tdes = log_Gl.trantable.all_tdes[i];
+      if (tdes != NULL && tdes->trid != NULL_TRANID && tdes->trid == trid)
+	{
+	  LSA_COPY (&tdes->head_lsa, log_lsa);
+	  tran_map.insert (std::pair<TRANID, LOG_TDES *> (trid, tdes));
```

Comment:
There are two issues with this line:
  - The tdes insterted into the map should not be the same as the tdes in log_Gl.trantable.all_tdes. Making any change to either tdes, in the map or in the transaction table, will modify the other one too. When tdes full compare is executed, the values are guaranteed to be the same. Create a new tdes and insert it into the map.
  - Find or insert should be executed, not insert. logtb_rv_find_allocate_tran_index may be called twice per transaction and the second insert may overwrite the first one. If there is a tdes in the map already, return the existing tdes.

```suggestion
	  tran_map.insert (std::pair<TRANID, LOG_TDES *> (trid, tdes));
```